src/hypothesis/strategies.py

```diff
@@ -836,31 +836,80 @@ def dm_func(denom):
 
 @cacheable
 @defines_strategy
-def decimals(
-    min_value=None, max_value=None, allow_nan=None, allow_infinity=None
-):
-    """Generates instances of decimals.Decimal.
+def decimals(min_value=None, max_value=None,
+             allow_nan=True, allow_infinity=None, places=None):
+    """Generates instances of decimals.Decimal, which may be:
 
-    - If min_value is not None, all values will be >= min_value.
-    - If max_value is not None, all values will be <= max_value.
-    - If min_value or max_value is not None, it is an error to enable
-      allow_nan.
-    - If both min_value and max_value are not None, it is an error to enable
-      allow_infinity.
+    - A finite rational number, between ``min_value`` and ``max_value``.
+    - Not a Number, if ``allow_nan`` is truthy.
+    - Positive or negative infinity, if ``max_value`` and ``min_value``
+      respectively are None, and ``allow_infinity`` is not False.  None means
+      "allow infinity, unless excluded by the min and max values".
 
-    Where not explicitly ruled out by the bounds, all of infinity, -infinity
-    and NaN are possible values generated by this strategy.
+    Note that where floats have one `NaN` value, Decimals have four: signed,
+    and either *quiet* or *signalling*.  See `the decimal module docs
+    <https://docs.python.org/3/library/decimal.html#special-values>`_ for
+    more information on special values.
+
+    If ``places`` is not None, all finite values drawn from the strategy will
+    have that number of digits after the decimal place.
 
     """
-    floats_strategy = floats(
-        min_value=min_value, max_value=max_value,
-        allow_nan=allow_nan, allow_infinity=allow_infinity)
-    fractions_strategy = fractions(min_value=min_value, max_value=max_value)
-
-    return (
-        floats_strategy.map(float_to_decimal) |
-        fractions_strategy.map(lambda f: Decimal(f.numerator) / f.denominator)
-    )
+    # Convert min_value and max_value to Decimal values, and validate args
+    check_valid_integer(places)
+    if places is not None and places < 0:
+        raise InvalidArgument('places=%r may not be negative' % places)
+
+    if min_value is not None:
+        min_value = Decimal(min_value)
+        if min_value.is_infinite() and min_value < 0:
```

Comment:
This looks funny if ```is_infinite()``` and ```is_finite()``` are contradictory.

In that case, what you've got is:

```
if A and B:
  // do stuff
elif A:
  // error
```

which could be:

```
if A:
  if B:
    // do stuff
  else:
    // error
```

If they are not contradictory, then it is is probably worth a comment.

But maybe conditional should be:

```
if min_value.is_infinite():
  if min_value == Decimal('-Infinity'):
    min_value = None
  else:
    raise ...
```

in that case, for greater precision.